Added contrast verification and readability checking to button.

jorgefilipecosta opened the pull request:
This PR aims to address part of https://github.com/WordPress/gutenberg/issues/2381, by adding a contrast checking following WCAG 2.0- AA (https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) to the button.

## Testing
Test some colors in button and verify that when contrast is high no message, in low contrast color pairs a message appear.
Verify that when background is darker than text, we suggest making background darker and/or text lighter and when text is darker we suggest making text darker and/or background lighter

## Screenshots (jpeg or gifs if applicable):
<img width="798" alt="screen shot 2017-11-08 at 19 50 32" src="https://user-images.githubusercontent.com/11271197/32571084-2967382a-c4be-11e7-8fbd-dd454d157834.png">

## Notes
This verification is implemented thanks to tinycolor2 lib. React-color (used in our color palette) already depends on it. This lib has no external dependencies. A new component was created that uses tinycolor2 to verify readability and then shows the correct warning message.

jorgefilipecosta committed:
Button block: Added contrast verification and readability checking to button.

This verification is implemented thanks to tinycolor2 lib. React-color (used in our color palette) already depends on it, this lib has no external dependencies. A new component was created that uses tinycolor2 to verify readability and then shows the correct warning message.

aaronjorbin commented:
Love where this is going.  Helping people make 

Should a state change or event of some sort when the contrast passing state changes? Thinking of the use case of a plugin that wants to block publishing of non-accessible content. 

I would love to see some tests of the new code, even though they would essentially be wrappers for testing `tinycolor2`. 

Also cc'ing @afercia so he sees this PR.

mcsf committed:
Button block: Refactor as a class

As an intermediate step before the next commit

mcsf committed:
Button block: Read fallback colors from DOM

mcsf commented:
Nice work!

**1.** Design-wise, I found it odd that ContrastChecker in the Button block should be rendered inside a PanelBody. I assume this is so that, visually, the notice appears between the two palettes. However, hierarchically, I'm not sure it's the best. I would simply place it as the next sibling of the `clear` ToggleControl, but I'll leave this up to other reviewers.

**2.** One rather glaring limitation I found with the Button block is that, if one is using a default color, ContrastChecker can't help the user. Here is an example with a default background color and a custom text color which, together, are clearly not readable:
 
<img width="1028" alt="screen shot 2017-10-24 at 23 46 52" src="https://user-images.githubusercontent.com/150562/31971969-31dd22fc-b916-11e7-95fa-07a7e80cdc28.png">

The default color is provided by `{style,editor}.scss` and so the block logic is unaware of the color. Because these styles are overridable by themes, it would be a bad idea to provide a default color hardcoded in JS.

One hack that I could conceive to work around this would require post-`render` calls to `window.getComputedStyle` to read the actual visible color of the button block. I pushed 9c99810843aa7925bfd55faafa9ddd26d64ddfa6 to clarify what I mean, but your best bet is probably to delete that commit (and its parent) from the branch.

mcsf commented:
Ignoring how technically broken(*) my commit is 😁 , one thing it highlights is that the default color pairing (`#ffffff` over `#33b3db`) doesn't pass `isReadable`, regardless of WCAG priority level. /cc @jasmussen 

(*): If the block had colors already set and the user then reverts them to default, it takes two renders until the contrast notice appears. This is because fallback colors aren't kept in `this.state` in order to avoid constant re-renders; this could perhaps be iterated on by saving fallback colors to component state but only ever setting them once: after we've just rendered **if** we didn't have colors in attributes — meaning the default style was indeed used in the DOM and we can read it. A probably flimsier alternative would be to create a temporary hidden button to read its computed style:

```js
const virtual = this.node.cloneNode();
virtual.style.backgroundColor = ''
virtual.style.display = 'none';
this.node.parentElement.appendChild( virtual );
getComputedStyle( virtual ).backgroundColor // <- grab it
virtual.parentElement.removeChild( virtual );
```

jasmussen commented:
Love where this is going. I also like that there's a simple and clear warning message showing up, rather than a for-on-developers-obtuse contrast ratio. It's a win for the regular user. Note that it seemed to take a while for the message to show up for me — it didn't seem to show up for me until I started tweaking the text color (background alone didn't make it show up). 

@mcsf 

> Ignoring how technically broken(*) my commit is 😁 , one thing it highlights is that the default color pairing (#ffffff over #33b3db) doesn't pass isReadable, regardless of WCAG priority level. /cc @jasmussen

Here's a sneak peek at a button redesign ;)

![screen shot 2017-10-27 at 09 43 33](https://user-images.githubusercontent.com/1204802/32093079-56f4c734-bafb-11e7-8bd4-a6742bb5a3db.png)

mcsf commented:
> it didn't seem to show up for me until I started tweaking the text color (background alone didn't make it show up)

That's my own fault, @jasmussen. :) If you `git checkout 5e02fb`, Jorge's initial commit, the UI responds immediately—with the caveat that it can't detect color contrast if one or both of the colors is the default one.

> Here's a sneak peek at a button redesign ;)

\o/

jorgefilipecosta committed:
Button block: Added fallback colors to component state.

This change should solve some refresh problems on contrast verification.

jorgefilipecosta commented:
Hi @aaronjorbin thank you for your sharing your thoughts. Regarding blocking, although that is not yet addressed in this PR, an idea I have in mind is adding to the contrast component an onFail property. Then parent components or other artifacts using their extensibility API (when ready) can make actions when contrast is failing like not allowing publishing.
Regarding testing I totally agree some tests should be added, they were not added yet because we are still prototyping but as soon as we have a clear direction I will add them.

Hi @mcsf thank you for your commits. I like the idea of using the computed styles, the only with this approach is that the color used on the editor may not be the final ones. Some themes change them. One option is adding the CSS of the theme and trying to read the final colors from there (may not be easy). The other option is using different messages for this case something like "Unless your theme uses different colors from the ones being used in the editor...". And the third option is keep things as they are, our message says "may be".

Hi @jasmussen, regarding the refresh problems I added a fix suggested by @mcs, now fallback colors are in the component state and the refresh problems should not happen.
I also changed the contrast verifier to the bottom of the colors and it is always visible even if colors are not.

The last improvements are available to test in the button component,

mcsf commented on blocks/library/button/index.js:
```diff
@@ -16,70 +17,106 @@ import BlockControls from '../../block-controls';
+	grabColors() {
+		const { background, text } = this.containers;
+		const { textColor, color } = this.props.attributes;
+		const { fallbackTextColorInState, fallbackBackgroundColorInState } = this.state;
```
When are these ever set? :)

We can make do with just the truthiness of `this.state.fallback*ColorInState`.

jorgefilipecosta replied:
Yes my logic had a bug, but the results looked like the expected in tests so I did not notice, it was solved :)

jorgefilipecosta commented:
Hi @mcsf, this was rebased and retested :) Could you have a new look and share your thoughts if this is ready or some change is needed?

jasmussen commented:
From a UI and functionality pov, this is 👍 👍  to me.

mcsf commented on blocks/contrast-checker/index.js:
```diff
@@ -0,0 +1,43 @@
+		msg = __( 'This color combination may be hard for people to read. Try using a darker background color and/or a brighter text color.' );
+	} else {
+		msg = __( 'This color combination may be hard for people to read. Try using a brighter background color and/or a darker text color.' );
+	}
```
Minor, but this is a good place for a multiline ternary const assignment.

mcsf reviewed (approved):
Looks good, thanks for working on this!
